src/main/java/com/google/sps/model/GmailClient.java

```diff
@@ -189,4 +202,30 @@ static String oneOfPhrasesInSubjectLineQuery(List<String> phrases) {
 
     return String.format("subject:(%s)", String.join(" OR ", phrases));
   }
+
+  /**
+   * Given a list of message headers, extract a single header with the specified name
+   *
+   * @param message a Gmail message object. Must be METADATA or FULL format
+   * @param headerName name of header that should be extracted
+   * @return first header in list with name headerName (this method does not handle duplicate
+   *     headers)
+   * @throws GmailMessageFormatException if the message does not contain the specified header (and
+   *     is thus the wrong format)
+   */
+  static MessagePartHeader extractHeader(Message message, String headerName) {
+    MessagePart payload = message.getPayload();
+    if (payload == null) {
+      throw new GmailMessageFormatException("No headers present!");
```

Comment:
It's also possible that payload.getHeaders() is empty, but you would return a different error. Perhaps make this error message more specific.

Reply:
Agreed - fixed!